_posts/2021-02-02-enso-prediction/enso-prediction.Rmd

```diff
@@ -0,0 +1,2410 @@
+  
+}
+
+valid_batch <- function(b) {
+  
+  output <- net(b$x$to(device = device))
+  
+  sst_output <- output[[1]]
+  sst_target <- b$y1$to(device = device)
+  
+  sst_loss <- nnf_mse_loss(sst_output[sst_target != 0], sst_target[sst_target != 0])
+  temp_loss <- nnf_mse_loss(output[[2]], b$y2$to(device = device))
+  nino_loss <- nnf_cross_entropy(output[[3]], b$y3$to(device = device))
+  
+  loss <-
+    lw_sst * sst_loss + lw_temp * temp_loss + lw_nino * nino_loss
+
+  list(sst_loss$item(),
+       temp_loss$item(),
+       nino_loss$item(),
+       loss$item())
+}
+
+for (epoch in 1:num_epochs) {
+  
+  net$train()
+  
+  train_loss_sst <- c()
+  train_loss_temp <- c()
+  train_loss_nino <- c()
+  train_loss <- c()
+
+  coro::loop(for (b in train_dl) {
+    losses <- train_batch(b)
+    train_loss_sst <- c(train_loss_sst, losses[[1]])
+    train_loss_temp <- c(train_loss_temp, losses[[2]])
+    train_loss_nino <- c(train_loss_nino, losses[[3]])
+    train_loss <- c(train_loss, losses[[4]])
+  })
+  
+  cat(
+    sprintf(
+      "\nEpoch %d, training: loss: %3.3f sst: %3.3f temp: %3.3f nino: %3.3f \n",
+      epoch, mean(train_loss), mean(train_loss_sst), mean(train_loss_temp), mean(train_loss_nino)
+    )
+  )
+  
+  net$eval()
+  
+  valid_loss_sst <- c()
+  valid_loss_temp <- c()
+  valid_loss_nino <- c()
+  valid_loss <- c()
+
+  coro::loop(for (b in valid_dl) {
+    losses <- valid_batch(b)
+    valid_loss_sst <- c(valid_loss_sst, losses[[1]])
+    valid_loss_temp <- c(valid_loss_temp, losses[[2]])
+    valid_loss_nino <- c(valid_loss_nino, losses[[3]])
+    valid_loss <- c(valid_loss, losses[[4]])
+    
+  })
+  
+  cat(
+    sprintf(
+      "\nEpoch %d, validation: loss: %3.3f sst: %3.3f temp: %3.3f nino: %3.3f \n",
+      epoch, mean(valid_loss), mean(valid_loss_sst), mean(valid_loss_temp), mean(valid_loss_nino)
+    )
+  )
+  
+  torch_save(net, paste0(
+    "model_", epoch, "_", round(mean(train_loss), 3), "_", round(mean(valid_loss), 3), ".pt"
+  ))
+  
+}
+
+```
+
+When I ran this, performance on the training set decreased in a not-too-fast, but continuous way, while validation set performance kept fluctuating. For reference, total (composite) losses looked like this:
+
+    Epoch     Training    Validation
+       
+       10        0.336         0.633
+       20        0.233         0.295
+       30        0.135         0.461
+       40        0.099         0.903
+       50        0.061         0.727
+
+Thinking of the size of the validation set - thirty-one years, or equivalently, 372 data points -- those fluctuations may not be all too surprising.
+
+# Predictions
+
+Now losses tend to be abstract; let's see what actually gets predicted. We obtain predictions for index values and phases like so ...
+
+```{r}
+net$eval()
+
+pred_index <- c()
+pred_phase <- c()
+
+coro::loop(for (b in valid_dl) {
+
+  output <- net(b$x$to(device = device))
+
+  pred_index <<- c(pred_index, output[[2]]$to(device = "cpu"))
```

Comment:
I think you don't need the `<<-`, `<-`should work

Reply:
oh yeah, will do!